File: crates/taflow-core/src/stream/time_series_rank.rs
//! Batch implementation for `time_series_rank`.

use super::rolling_rank::rolling_rank;
use crate::error::TaResult;

/// Canonical WorldQuant name for the existing causal rolling-rank state.
pub type TimeSeriesRank = super::RollingRank;

/// WorldQuant Alpha101 time-series rank: the rank of the current value within
/// the trailing `d`-bar window as a fraction in `(0, 1]`. Shares the rolling
/// Compute the time series rank result for the supplied aligned series.
///
/// # Parameters
///
/// * `input` - Input series or configuration value.
/// * `timeperiod` - Input series or configuration value.
///
/// # Returns
///
/// An aligned result with TA-Lib-compatible validation and warm-up values.
pub fn time_series_rank(input: &[f64], timeperiod: usize) -> TaResult<Vec<f64>> {
    rolling_rank(input, timeperiod)
}

File: python/taflow/indicators/time_series_rank.py
"""Canonical native-backed time-series-rank adapter."""

from typing import Any
import numpy as np
from .._adapter_protocol import adapter_length
from .._native import RollingRank as _Native
from .._series import as_float64_series


class TimeSeriesRank:
    """WorldQuant rank of each value within a trailing window."""

    def __init__(self, timeperiod: int = 14) -> None:
        self._state = _Native(int(timeperiod))

    def append(self, input: float) -> "TimeSeriesRank":
        self._state.append(float(input))
        return self

    def extend(self, input: Any) -> "TimeSeriesRank":
        values = as_float64_series(input)
        self._state.extend(values)
        return self

    def compute(self) -> np.ndarray:
        return self._state.compute()

    @property
    def value(self) -> float | None:
        return self._state.value

    def reset(self) -> "TimeSeriesRank":
        self._state.reset()
        return self

    def __len__(self) -> int:
        return adapter_length(self)
